client/ctclient/ctclient.go

```diff
@@ -138,41 +138,28 @@ func getEntries(ctx context.Context, logClient *client.LogClient) {
 		switch ts.EntryType {
 		case ct.X509LogEntryType:
 			fmt.Printf("X.509 certificate:\n")
-			showCert(*ts.X509Entry)
+			showCert(entry.X509Cert.Raw)
 		case ct.PrecertLogEntryType:
-			fmt.Printf("pre-certificate from issuer with keyhash %x:\n", ts.PrecertEntry.IssuerKeyHash)
-			showTBSCert(ts.PrecertEntry.TBSCertificate)
+			fmt.Printf("pre-certificate from issuer with keyhash %x:\n", entry.Precert.IssuerKeyHash)
+			showCert(entry.Precert.Raw)
 		default:
 			log.Fatalf("Unhandled log entry type %d", entry.Leaf.TimestampedEntry.EntryType)
 		}
 	}
 }
 
-func showCert(cert ct.ASN1Cert) {
-	if *textOut {
-		c, err := x509.ParseCertificate(cert.Data)
-		if err != nil {
-			log.Printf("Error parsing certificate: %q", err.Error())
-			return
-		}
-		fmt.Printf("%s\n", x509util.CertificateToString(c))
-	} else {
-		if err := pem.Encode(os.Stdout, &pem.Block{Type: "CERTIFICATE", Bytes: cert.Data}); err != nil {
+func showCert(cert []byte) {
+	if *pemOut {
+		if err := pem.Encode(os.Stdout, &pem.Block{Type: "CERTIFICATE", Bytes: cert}); err != nil {
 			log.Printf("Failed to PEM encode cert: %q", err.Error())
 		}
-	}
-}
-
-func showTBSCert(tbs []byte) {
-	if *textOut {
-		c, err := x509.ParseTBSCertificate(tbs)
+	} else {
+		c, err := x509.ParseCertificate(cert)
```

Comment:
Like this? It feels weird to me; maybe I'm misunderstanding.
```
func showRawCert(cert []byte) {
	if *textOut {
		c, err := x509.ParseCertificate(cert)
		if err != nil {
			log.Printf("Error parsing certificate: %q", err.Error())
			return
		}
		showParsedCert(c)
	} else {
		if err := pem.Encode(os.Stdout, &pem.Block{Type: "CERTIFICATE", Bytes: cert}); err != nil {
			log.Printf("Failed to PEM encode cert: %q", err.Error())
		}
	}
}

func showParsedCert(cert *x509.Certificate) {
	if *textOut {
		fmt.Printf("%s\n", x509util.CertificateToString(cert))
	} else {
		showRawCert(cert.Raw)
	}
}
```

Reply:
Yeah, the mutual recursion thing does turn out a bit weird, doesn't it?

I still like the idea of not re-parsing (and also btw of taking parameter types that imply structure, like `ct.ASN1Cert` rather than `[]byte`), so something like the following might less odd:
```go
func showRawCert(cert ct.ASN1Cert) {
	if *textOut {
		c, err := x509.ParseCertificate(cert.Data)
		if err != nil {
			log.Printf("Error parsing certificate: %q", err.Error())
			return
		}
		showParsedCert(c)
	} else {
		showPEMData(cert.Data)
	}
}

func showParsedCert(cert *x509.Certificate) {
	if *textOut {
		fmt.Printf("%s\n", x509util.CertificateToString(cert))
	} else {
		showPEMData(cert.Raw)
	}
}

func showPEMData(data []byte) {
	if err := pem.Encode(os.Stdout, &pem.Block{Type: "CERTIFICATE", Bytes: data}); err != nil {
		log.Printf("Failed to PEM encode data: %q", err.Error())
	}
}
```